pandas/core/frame.py

```diff
@@ -3591,14 +3594,27 @@ def transpose(self, *args, copy: bool = False) -> DataFrame:
             and dtypes
             and isinstance(dtypes[0], ExtensionDtype)
         ):
-            # We have EAs with the same dtype. We can preserve that dtype in transpose.
-            dtype = dtypes[0]
-            arr_type = dtype.construct_array_type()
-            values = self.values
+            if isinstance(dtypes[0], BaseMaskedDtype):
+                # We have masked arrays with the same dtype. We can transpose faster.
+                from pandas.core.arrays.masked import transpose_homogenous_masked_arrays
+
+                if isinstance(self._mgr, ArrayManager):
+                    masked_arrays = self._mgr.arrays
+                else:
+                    masked_arrays = [blk.values for blk in self._mgr.blocks]
```

Comment:
might be safer to use iter_column_arrays?  this might mess up if blocks get shuffled in a weird order

Reply:
I've changed it.